hax/test/integration/test_motr.py

```diff
@@ -71,6 +73,109 @@ def fn(trace: Invocation) -> bool:
     return fn
 
 
+def node_failed() -> TraceMatcher:
+    '''
+    Returns true if the second argument is a array to pointer to HaNoteStruct
+    and it contains 5 elements and that structure brings M0_NC_FAILED state.
+    In case of node failure we should see 5 elements(1 IO service, 1 node fid,
+    1 encl fid, 2 ctrls) in array.
+
+    It makes sense to use this matcher together with
+    `tr_method('ha_broadcast')`
+    '''
+    def fn(trace: Invocation) -> bool:
+        if len(trace.args) < 2:
+            return False
+        notes = trace.args[1]
+        if (not isinstance(notes, ctypes.Array) or
+                len(notes) != 5):
+            return False
+
+        fids = [FidStruct(0x7200000000000001, 0x15),
+                FidStruct(0x6e00000000000001, 0x3),
+                FidStruct(0x6500000000000001, 0x4),
+                FidStruct(0x6300000000000001, 0x5),
+                FidStruct(0x6300000000000001, 0x6)]
+
+        for _, note in enumerate(notes):
```

Comment:
If the call was invoked with one and the same fid duplicated multiple times, this matcher will still return true.

In other words, if the HaNoteStruct erroneously contains `[FidStruct(0x6300000000000001, 0x6), FidStruct(0x6300000000000001, 0x6), FidStruct(0x6300000000000001, 0x6)]`, this matcher will return True.

Reply:
Done